### PR TITLE
Rewrite cache loader to async framework

### PR DESCRIPTION
**Pull request by godexsoft**
First rewrite with the async framework. :rage3:

### CODE REVIEW

File: src/etl/CacheLoader.hpp [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        ASSERT(not cache_.get().isFull(), "Cache must not be full. seq = {}", seq);
+
+        if (settings_.isDisabled()) {
+            cache_.get().setDisabled();
+            LOG(log_.warn()) << "Cache is disabled. Not loading";
+            return;
+        }
+
+        auto const provider = CursorProviderType{backend_, settings_.numCacheDiffs};
+        loader_ = std::make_unique<CacheLoaderType>(
```
Comment (kuznetsss) [inline src/etl/CacheLoader.hpp:92]: Is it ok that if `load()` is called 2 times with different sequences `loader_` will be recreated?

File: src/etl/CacheLoaderSettings.hpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+#include <cstddef>
+
+namespace etl {
+
+/**
+ * @brief Settings for the cache loader
+ */
+struct CacheLoaderSettings {
+    enum class LoadStyle { ASYNC, SYNC, NOT_AT_ALL };
```
Comment (kuznetsss) [inline src/etl/CacheLoaderSettings.hpp]: Maybe `NONE` instead of `NOT_AT_ALL`?

File: src/etl/CacheLoaderSettings.hpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+    size_t numCacheDiffs = 32;       /**< number of diffs to use to generate cursors */
+    size_t numCacheMarkers = 48;     /**< number of markers to use at one time to traverse the ledger */
+    size_t cachePageFetchSize = 512; /**< number of ledger objects to fetch concurrently per marker */
+    size_t numThreads = 2;           /**< number of threads to use for loading cache */
+
+    LoadStyle loadStyle = LoadStyle::ASYNC; /**< how to load the cache */
+
+    int
```
Comment (kuznetsss) [inline src/etl/CacheLoaderSettings.hpp]: Why `int`?

File: src/etl/CacheLoaderSettings.hpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    operator<=>(CacheLoaderSettings const&) const = default;
```
Comment (kuznetsss) [inline src/etl/CacheLoaderSettings.hpp:42]: Maybe `operator==() const = default` would be enough? Do you need operators `<`,`>` and others?

File: src/etl/impl/CacheLoader.hpp [hunk 5]
```diff
@@ -1,7 +1,7 @@
 //------------------------------------------------------------------------------
 /*
     This file is part of clio: https://github.com/XRPLF/clio
-    Copyright (c) 2023, the clio developers.
+    Copyright (c) 2024, the clio developers.
```
Comment (kuznetsss) [inline src/etl/impl/CacheLoader.hpp]: Are you sure? I thought this year is the year of the first appearance of this file.

File: src/etl/impl/CacheLoader.hpp [hunk 6]
```diff
@@ -20,423 +20,133 @@
...
-            LOG(log_.info()) << "Finished downloading ledger from clio node. ip = " << ip;
-
-            cache_.get().setFull();
-            return true;
-        } catch (std::exception const& e) {
-            LOG(log_.error()) << "Encountered exception : " << e.what() << " - ip = " << ip;
-            return false;
-        }
+        LOG(log_.info()) << "Loading cache. Num cursors = " << queue_.size();
+        for ([[maybe_unused]] auto taskId : vs::iota(0u, numCacheMarkers))
```
Comment (kuznetsss) [inline src/etl/impl/CacheLoader.hpp:108]: `tasks_.reserve(numCacheMarkers)`

File: src/etl/impl/CursorProvider.hpp [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+            std::unique(
+                std::begin(diffs), std::end(diffs), [](auto const& a, auto const& b) { return a.key == b.key; }
+            ),
+            std::end(diffs)
+        );
+
+        std::vector<ripple::uint256> cursors{data::firstKey};
+        rg::copy(
+            diffs                                                                   //
+                | vs::filter([](auto const& obj) { return not obj.blob.empty(); })  //
```
Comment (kuznetsss) [inline src/etl/impl/CursorProvider.hpp:89]: `//` without a comment

File: src/util/async/impl/ErasedOperation.hpp [hunk 8]
```diff
@@ -61,8 +61,8 @@ class ErasedOperation {
     }
 
     /**
-     * @brief Request the operation to be stopped as soon as possible.
-     * @throw std::logic_error if the erased operation is non-stoppable
+     * @brief Request the operation to be stopped as soon as possible
+     * Note: ASSERTs if the operation is not stoppable
```
Comment (kuznetsss) [inline src/util/async/impl/ErasedOperation.hpp]: `@note`

File: unittests/etl/CacheLoaderTests.cpp [hunk 9]
```diff
@@ -1,7 +1,7 @@
 //------------------------------------------------------------------------------
 /*
     This file is part of clio: https://github.com/XRPLF/clio
-    Copyright (c) 2023, the clio developers.
+    Copyright (c) 2024, the clio developers.
```
Comment (kuznetsss) [inline unittests/etl/CacheLoaderTests.cpp]: Also I think year shouldn't be changed.

File: unittests/etl/CursorProviderTests.cpp [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    auto const provider = etl::impl::CursorProvider{backend, 0};
+
+    ON_CALL(*backend, fetchLedgerDiff(_, _)).WillByDefault(Return(diffs));
+    EXPECT_CALL(*backend, fetchLedgerDiff(_, _)).Times(0);
+
+    auto const cursors = provider.getCursors(SEQ);
+
+    EXPECT_EQ(cursors.front().start, firstKey);
+    EXPECT_EQ(cursors.back().end, lastKey);
+    EXPECT_EQ(cursors.size(), 1);
```
Comment (kuznetsss) [inline unittests/etl/CursorProviderTests.cpp]: Better to check the size using `ASSERT_EQ()` before calling `front()` and `back()`

File: unittests/etl/CursorProviderTests.cpp [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+{
+    auto const numDiffs = GetParam();
+    auto const diffs = diffProvider.getLatestDiff();
+    auto const provider = etl::impl::CursorProvider{backend, numDiffs};
+
+    ON_CALL(*backend, fetchLedgerDiff(_, _)).WillByDefault(Return(diffs));
+    EXPECT_CALL(*backend, fetchLedgerDiff(_, _)).Times(numDiffs);
+
+    auto const cursors = provider.getCursors(SEQ);
+    EXPECT_EQ(cursors.size(), diffs.size() + 1);
```
Comment (kuznetsss) [inline unittests/etl/CursorProviderTests.cpp]: Better `ASSERT_EQ()`

**Review by kuznetsss — COMMENTED** [review]
The code looks much better now! Just small comments

File: src/etl/CacheLoader.hpp (hunk 1 above)
Comment (godexsoft) [reply src/etl/CacheLoader.hpp:92]: Yes, that's the idea. The previous one would get cancelled. But tbh., at that point we already have some data in the cache. As long as it's not `full` it should be ok to start over.

File: src/etl/CacheLoaderSettings.hpp (hunk 3 above)
Comment (godexsoft) [reply src/etl/CacheLoaderSettings.hpp]: My bad, should be `auto` or one of the custom comparison categories, will change

File: src/etl/CacheLoaderSettings.hpp (hunk 4 above)
Comment (godexsoft) [reply src/etl/CacheLoaderSettings.hpp:42]: Meh, why use `==` if you have spaceship :)

File: src/etl/impl/CursorProvider.hpp (hunk 7 above)
Comment (godexsoft) [reply src/etl/impl/CursorProvider.hpp:89]: Yea, that's to align the pipes properly. The alternative is to use `// clang-format off` etc. - too much extra comments then

**Review by kuznetsss — APPROVED** [review]

**Pull request merged**